### PR TITLE
Adds MSP to CRSF over TCP (WIFI) feature

### PR DESCRIPTION
**Pull request by AlessandroAU**
Look ma, no cables!

Adds three new abstraction libraries:

msp2crsf
crsf2msp
tcpsocket

The first two libraries convert raw MSP() frames between MSP encapsulated CRSF frames and back. They support MSPv1 MSPV1Jumbo and MSPv2. Frames are automatically chunked into segments and reassembled as is required by the CRSF protocol. Buffers are sized to handle 512B MSP frames. tcpsocket provides an abstraction over the ESPAsyncTCP to buffer in/out frames. This can also be used to send arbitary MSP frames down the CRSF pipe, and we should try to migrate some of the CRSF CMD frames to MSP frames in the future. Eg the VTX commands.

The main immediate application is allowing betaflight configurator to be used wirelessly through any wifi-capable RX ExpressLRS receiver. Currently listens on port 5761 which is default for the TCP socket for BF configurator. The users needs to either type in the address of the receiver on the local network when on home wifi, or enter tcp://10.0.0.1 when in hotspot mode. 

image
Demo:
ezgif-3-9741986ee7

Has also been tested to work with betaflight configurator on Andriod. 

Known issues:
CLI doesn't work (there is currently no way encapsulate MSP frames to CLI)
Presets don't work (requires CLI)
Setting the target address as tcp://elrs_rx in betaflight configurator doesn't work as it doesn't try to resolve hostnames (upstream issue)

### CODE REVIEW

**Comment by CapnBry** [comment]
This definitely worked when I was testing it before the PR, but if I switch over to the tcp2msp branch now I get all timeouts in the configurator with BF 4.3 RC2 (which theoretically has your fix for the bigger CRSF MSP timeouts and buffer size)
image

Some pages will eventually load, but mostly it just gets stuck. When running on my home network instead of just in AP mode, I can't even usually get a real connection through the configurator, it times out during the handshake:
```
2022-01-14 @15:48:10 -- Serial port successfully opened with ID: 4
2022-01-14 @15:48:11 -- MultiWii API version: 1.44.0
2022-01-14 @15:48:14 -- Flight controller info, identifier: BTFL, version: 4.3.0
2022-01-14 @15:48:20 -- No configuration received within 10 seconds, communication failed
2022-01-14 @15:48:20 -- Serial port successfully closed
```

**Comment by pkendall64** [comment]
@CapnBry make sure you do not have -DDEBUG_LOG otherwise it causes problems with the link.
I had this problem and it took me hours before I realised my mistake!

**Comment by CapnBry** [comment]
> @CapnBry make sure you do not have -DDEBUG_LOG otherwise it causes problems with the link. I had this problem and it took me hours before I realised my mistake!

Ah crepes, PK, you're right. I did build it with DEBUG_LOG on like a damn fool. Works perfectly like it did when I tried it before the PR. Thank you!

**Comment by AlessandroAU** [comment]
Ah sorry guys, yeah I should have mentioned! Betaflight does not like the debug output getting spit over serial.

**Comment by pkendall64** [comment]
> Ah sorry guys, yeah I should have mentioned! Betaflight does not like the debug output getting spit over serial.

I wonder if on the RXes we could dual use the GPIO0 (boot pad) to be a soft serial for debug output.

**Comment by AlessandroAU** [comment]
Not a bad idea, I've also gotten debug via websockets to work, so that's another option.

**Comment by pkendall64** [comment]
> Not a bad idea, I've also gotten debug via websockets to work, so that's another option.

I like the websockets, but the problem is that you can't really run wifi and the radio on the RX at the same time!
I think the wifi is just too hungry to allow the radio to work reliably.

**Comment by AlessandroAU** [comment]
Yeah, good point, but I don't think we tested that since you changed over to the async library. We should try it again.

**Comment by pkendall64** [comment]
> Yeah, good point, but I don't think we tested that since you changed over to the async library. We should try it again.

Ah, true. Perhaps I'll give that go later then.

**Comment by RipperGOT** [comment]
Can work with Inav?
The Speedybee app only connects to the port of their own device, right?

**Comment by AlessandroAU** [comment]
This doesn't support speedybee, and probably can't. Inav should work too, but there was some upstream changes needed in BF to fix MSP-CRSF encapsulation. Inav will probably be the same

File: src/lib/CRSF/CRSF.cpp [hunk 1]
```diff
@@ -860,26 +865,39 @@ bool CRSF::UARTwdt()
...
 bool CRSF::RXhandleUARTout()
 {
+    bool retVal = false;
 #if !defined(CRSF_RCVR_NO_SERIAL)
-    uint8_t peekVal = SerialOutFIFO.peek(); // check if we have data in the output FIFO that needs to be written
-    if (peekVal > 0)
-    {
-        if (SerialOutFIFO.size() > (peekVal))
+    // don't write more than 128 bytes at a time to avoid RX buffer overflow
+    #define maxBytesPerCall 128
```
Comment (pkendall64) [inline src/lib/CRSF/CRSF.cpp:871]: prefer 
```cpp
const int maxBytesPerCall = 128;
```
so it doesn't leak outside the function

File: src/lib/CRSF/CRSF.h [hunk 2]
```diff
@@ -3,6 +3,10 @@
 
 #include "targets.h"
 #include "crsf_protocol.h"
+#if defined(PLATFORM_ESP8266) && defined(CRSF_RX_MODULE)
```
Comment (pkendall64) [inline src/lib/CRSF/CRSF.h:6]: ```#if defined(PLATFORM_ESP8266) && defined(CRSF_RX_MODULE) && defined(USE_MSP_WIFI)```

File: src/lib/CRSF/CRSF.h [hunk 3]
```diff
@@ -31,7 +35,13 @@ class CRSF
     }
 
     CRSF(Stream &dev) : _dev(&dev) {}
+
+    #if defined(PLATFORM_ESP8266)
```
Comment (pkendall64) [inline src/lib/CRSF/CRSF.h:39]: ```#if defined(PLATFORM_ESP8266) && defined(USE_MSP_WIFI)```

File: src/lib/TCPSOCKET/tcpsocket.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
+#pragma once
+#if defined(USE_MSP_WIFI)
```
Comment (pkendall64) [inline src/lib/TCPSOCKET/tcpsocket.h:2]: ```#if defined(PLATFORM_ESP8266) && defined(USE_MSP_WIFI)```

File: src/lib/WIFI/devWIFI.cpp [hunk 5]
```diff
@@ -617,6 +624,9 @@ static void startServices()
 
   servicesStarted = true;
   DBGLN("HTTPUpdateServer ready! Open http://%s.local in your browser", wifi_hostname);
+  #if defined(USE_MSP_WIFI)
```
Comment (pkendall64) [inline src/lib/WIFI/devWIFI.cpp:627]: `  #if defined(USE_MSP_WIFI) && defined(TARGET_RX)`

File: src/lib/WIFI/devWIFI.cpp [hunk 6]
```diff
@@ -59,6 +59,13 @@ static IPAddress netMsk(255, 255, 255, 0);
 static DNSServer dnsServer;
 static IPAddress ipAddress;
 
+#if defined(USE_MSP_WIFI)  //MSP2WIFI in enabled only for RX only at the moment
```
Comment (pkendall64) [inline src/lib/WIFI/devWIFI.cpp:62]: `  #if defined(USE_MSP_WIFI) && defined(TARGET_RX)`

**Review by pkendall64 — COMMENTED** [review]
Basically just making the #ifdef's consistent

**Comment by yanwenjun** [comment]
> @CapnBry make sure you do not have -DDEBUG_LOG otherwise it causes problems with the link. I had this problem and it took me hours before I realised my mistake!

I am a fool novice, how can I turn off DEBUG_LOG?

**Comment by pkendall64** [comment]
> > @CapnBry make sure you do not have -DDEBUG_LOG otherwise it causes problems with the link. I had this problem and it took me hours before I realised my mistake!
> 
> I am a fool novice, how can I turn off DEBUG_LOG?

If you are using configurator then I think it's off already.

**Comment by yanwenjun** [comment]
Can this function also be achieved by connecting to the WIFI of the ELRS TX?

**Review by pkendall64 — APPROVED** [review]

**Comment by CapnBry** [comment]
> Can this function also be achieved by connecting to the WIFI of the ELRS TX?

No, that would require a whole lot more code to make happen, and the performance would be absolutely terrible. The configuration connection expects a pretty high bandwidth and low latency connection since it normally runs over serial. Running it over an extremely high latency link at ~5kbit throughput will cause Betaflight Configurator to timeout repeatedly and make a lot of redundant requests that will be discarded.

There would be a lot of work needed in BF Configurator to be tolerant of a low bandwidth high latency connection. It is just a mountain of work for both projects, and the user experience is going to still be pretty awful due to the link speed limitations like trying to stream full resolution video over a dialup modem. There are plenty of ESP-based ELRS RX that will support this though, so it isn't too much to ask of the user to just spend $15 on a new RX if they want this feature instead of getting it for free on old R9 hardware + TX MSP passthrough and then complaining about how terribly it works.

**Review by CapnBry — APPROVED** [review]

**Comment by Diehard-FPV** [comment]
image
I wonder what does it mean?Emm...may be I'm also a fool novice

**Comment by Diehard-FPV** [comment]
> 这不支持 speedybee，也可能不支持。Inav也应该可以工作，但是BF需要一些上游更改来修复MSP-CRSF封装。伊纳夫可能是一样的

no, i have tried connecting INAV & Emuflight, and it appears that it doesn't support neither of them.

**Comment by Diehard-FPV** [comment]
> This doesn't support speedybee, and probably can't. Inav should work too, but there was some upstream changes needed in BF to fix MSP-CRSF encapsulation. Inav will probably be the same

no, i have tried connecting INAV & Emuflight, and it appears that it doesn't support neither of them.

**Comment by reednoel4u** [comment]
I using Expresslrs Configulator for update firmware!
Please shơ me hom to adds three new abstraction libraries:
msp2crsf
crsf2msp
tcpsocket
Thanks

**Comment by pkendall64** [comment]
> I using Expresslrs Configulator for update firmware! Please show me how to adds three new abstraction libraries: msp2crsf crsf2msp tcpsocket Thanks

This code is merged into the `master` branch. If you are wishing to *test*, then you could flash master to both your TX module and an RX.

**Comment by pkendall64** [comment]
> > > I using Expresslrs Configulator for update firmware! Please show me how to adds three new abstraction libraries: msp2crsf crsf2msp tcpsocket Thanks
> > 
> > 
> > This code is merged into the `master` branch. If you are wishing to _test_, then you could flash master to both your TX module and an RX.
> > Could you please help me hoi to Flash master! i don't know so much this application, I just using flash fw with Expresslrs Configulator!

If you don't know how to flash `master` then I think you'd be best sticking to a release version i.e. 2.3.0 and wait till the devs release a new version with MSP over WIFI included.

If you are really keen to try this feature, then I suggest you ask on #help-and-support on the discord server.

**Comment by reednoel4u** [comment]
Thanks.
But I always to Flash new version (now 2.3.0).
I think, that waiting the Dev addition!

**Comment by howels** [comment]
> Thanks. But I always to Flash new version (now 2.3.0). I think, that waiting the Dev addition!

https://discord.gg/c93u2Q2h

**Comment by reednoel4u** [comment]
Thanks, I joint to discord . and me waiting for the nên version

**Comment by stuffshit** [comment]
@reednoel4u did you managed to run it?

**Comment by reednoel4u** [comment]
Not yet, I waitìng for the new version, that be add labrary tcp

**Comment by stuffshit** [comment]
hmmm. I'm using 2.4.0 firmware version with betafpv nano rx and betaflight can't connect.

**Comment by pkendall64** [comment]
This feature will not be available until release 3.0

**Comment by stuffshit** [comment]
Is it available in any other branch so I can merge it in my own?
> This code is merged into the `master` branch. If you are wishing to _test_, then you could flash master to both your TX module and an RX.

**Comment by stuffshit** [comment]
Is it available in any other branch so I can merge it in my own?

> This code is merged into the `master` branch. If you are wishing to _test_, then you could flash master to both your TX module and an RX.

**Comment by pkendall64** [comment]
> Is it available in any other branch so I can merge it in my own?
> 
> > This code is merged into the `master` branch. If you are wishing to _test_, then you could flash master to both your TX module and an RX.

No it's not on any other branch. All the changed files are in this PR, so you could try to put them into a 2.4 if you want to try it for yourself.

**Comment by stuffshit** [comment]
> > Is it available in any other branch so I can merge it in my own?
> > > This code is merged into the `master` branch. If you are wishing to _test_, then you could flash master to both your TX module and an RX.
> 
> No it's not on any other branch. All the changed files are in this PR, so you could try to put them into a 2.4 if you want to try it for yourself.

Good to know. Thanks

**Comment by stuffshit** [comment]
But as I can see this PR is already marked as merged

**Comment by pkendall64** [comment]
> But as I can see this PR is already marked as merged

Yes, to the master branch, which will become 3.0 when we are ready.

**Comment by stuffshit** [comment]
> Yes, to the master branch, which will become 3.0 when we are ready.
So flashing rx with the master branch should make this feature available, right?

**Comment by howels** [comment]
> > Yes, to the master branch, which will become 3.0 when we are ready.
> 
> So flashing rx with the master branch should make this feature available, right?

You need to flash TX also or TX won't talk to the RX.

**Comment by stuffshit** [comment]
> > > Yes, to the master branch, which will become 3.0 when we are ready.
> > 
> > 
> > So flashing rx with the master branch should make this feature available, right?
> 
> You need to flash TX also or TX won't talk to the RX. At least I think so, OTA protocol has had some updates for 3.0 and I think it breaks compatibility with 2.x

I guess TX is not needed in order this feature to work, but good to know it also have to be updated

**Comment by AVIC621** [comment]
Need some help!   master branch can not for build


image

**Comment by slmnow** [comment]
Great work!  Hoping a mavlink2crsf version for ardupilot is coming.

**Comment by linroid** [comment]
@AVIC621 From the error logs, I think you're compile __tx__ firmware through __rx__ config, please select the tx config

**Comment by fa-lk** [comment]
What would be needed to get this also working with INAV?

**Comment by RomanLut** [comment]
Lost few hours intill I noticed that "Telemetry output" is not enabled in "Receiver tab" in Betaflight configurator in my ad-hoc setup...

**Comment by CREASE-gum-EAR** [comment]
What's the baud rate of this serial over wifi connection? I use it for INAV configuration on my big planes and it can be a little slow. But that may be just wifi interference with such small wifi antennas. I was thinking of trying Airport for configuring at the bench. But is Wifi still wayyyy faster than Airport? This PR appears to be the only official documentation of this feature. Maybe after this I'll commit something to the docs.

**Comment by RomanLut** [comment]
> I use it for INAV configuration on my big planes and it can be a little slow.

This feature does not work with inav. Inav is missing mspV2 support in mspOverCrsf implementation. Inav configurator is sending mspv2 packets.

For inav you have to look at esp-link or bronebridge connected to msp uart.

**Comment by howels** [comment]
> What's the baud rate of this serial over wifi connection? I use it for INAV configuration on my big planes and it can be a little slow. But that may be just wifi interference with such small wifi antennas. I was thinking of trying Airport for configuring at the bench. But is Wifi still wayyyy faster than Airport? This PR appears to be the only official documentation of this feature. Maybe after this I'll commit something to the docs.

Slightly slower with eLRS than Crossifre and Tracer, so there was some modification recently in Betaflight to check the serial of the CRSF device - eLRS devices all have the serial EXPRESSLRS which can be used to tune throughput and avoid dropped MSP frames.
iNav should look at this - the RX uses mDNS to advertise itself so it's very friendly to autodetect on wifi

**Comment by psfales** [comment]
I was intrigued by this and noticed that the original comment says "Has also been tested to work with betaflight configurator on Andriod. [sic]"   Can anyone shed some light on what Android app works with this?   It works very well on my computer at home, but it would be great if could be used at the field with an Android app!

**Comment by klmn1k** [comment]
Can I flash firmware with it?

**Comment by pkendall64** [comment]
> Can I flash firmware with it?

It's built in to all V3 firmwares

**Comment by klmn1k** [comment]
I mean Betaflight. Can I update betaflight using this? Via wifi?

**Comment by klmn1k** [comment]
> > Can I flash firmware with it?
> 
> It's built in to all V3 firmwares

I mean Betaflight. Can I update betaflight using this? Via wifi?

**Comment by pkendall64** [comment]
> > > Can I flash firmware with it?
> > 
> > 
> > It's built in to all V3 firmwares
> 
> I mean Betaflight. Can I update betaflight using this? Via wifi?

No, you cannot update betaflight firmware via the wifi connection.

**Comment by bastian2001** [comment]
Hi there, and sorry for reviving such an old PR...
I know we all hate MSP Jumbo frames, but I think there might be something wrong in the implementation. From [the iNav docs](https://github.com/iNavFlight/inav/wiki/MSP-V2#v1-structure) it seems like in a V1 Jumbo frame, the length bytes are the indices 4 and 5, but in [msp2crsf.cpp, line 120](https://github.com/ExpressLRS/ExpressLRS/blob/7684347ee697b3fa4318f99a02b6ae6f5d703307/src/lib/CRSF2MSP/msp2crsf.cpp#L120-L121) bytes 5 and 6 are used. Similarly, [here](https://github.com/ExpressLRS/ExpressLRS/blob/7684347ee697b3fa4318f99a02b6ae6f5d703307/src/lib/CRSF2MSP/crsf2msp.cpp#L153-L154) bytes 2 and 3 are used, as opposed to 1 and 2 (byte 0 is 255, followed by the actual length in bytes 1 and 2). Am I missing something here?

**Comment by pkendall64** [comment]
> Am I missing something here?

Yes, MSP V1 Jumbo frames have the payload_size at index 3 set to 255, and the actual size is in the payload at index 5 & 6. Specified in the docs that you link to!

**Comment by bastian2001** [comment]
> is in the payload
Ah I totally misread that, thanks. I always thought that the length is set to 255, and then it's immediately followed by the actual size (i.e. before the command).

**Pull request merged**